tests/e2e/e2e_test.go

```diff
@@ -186,65 +186,167 @@ func (s *IntegrationTestSuite) CheckBalance(node *chain.NodeConfig, addr, denom
+			s.Require().True(balancesAfter.AmountOf("stake").Equal(balancesBefore.AmountOf("stake")))
+			s.Require().True(balancesAfter.AmountOf("e2e-default-feetoken").Equal(balancesBefore.AmountOf("e2e-default-feetoken").Sub(defaultFeePerTx)))
+			if assertUion {
+				s.Require().True(balancesAfter.AmountOf("uion").Equal(balancesBefore.AmountOf("uion")))
+			}
+			if assertUosmo {
+				s.Require().True(balancesAfter.AmountOf("uosmo").Equal(balancesBefore.AmountOf("uosmo")))
+			}
+		}
+
+		// calculateFeeGrowthGlobal calculates fee growth global per unit of virtual liquidity based on swap parameters:
+		// amountIn - amount being swapped
+		// swapFee - pool's swap fee
+		// poolLiquidity - current pool liquidity
+		calculateFeeGrowthGlobal = func(amountIn, swapFee, poolLiquidity sdk.Dec) sdk.Dec {
+			// First we get total fee charge for the swap (ΔY * swapFee)
+			feeChargeTotal := amountIn.Mul(swapFee)
+
+			// Calculating fee growth global (dividing by pool liquidity to find fee growth per unit of virtual liquidity)
+			feeGrowthGlobal := feeChargeTotal.Quo(poolLiquidity)
+			return feeGrowthGlobal
+		}
+
+		// calculateFeeGrowthInside calculates fee growth inside range per unit of virtual liquidity
+		// feeGrowthGlobal - global fee growth per unit of virtual liquidity
+		// feeGrowthBelow - fee growth below lower tick
+		// feeGrowthAbove - fee growth above upper tick
+		// Formula: feeGrowthGlobal - feeGrowthBelowLowerTick - feeGrowthAboveUpperTick
+		calculateFeeGrowthInside = func(feeGrowthGlobal, feeGrowthBelow, feeGrowthAbove sdk.Dec) sdk.Dec {
+			return feeGrowthGlobal.Sub(feeGrowthBelow).Sub(feeGrowthAbove)
+		}
+
+		// Helper function for calculating uncollected fees since the time that feeGrowthInsideLast corresponds to
+		// positionLiquidity - current position liquidity
+		// feeGrowthBelow - fee growth below lower tick
+		// feeGrowthAbove - fee growth above upper tick
+		// feeGrowthInsideLast - amount of fee growth inside range at the time from which we want to calculate the amount of uncollected fees
+		// feeGrowthGlobal - variable for tracking global fee growth
+		calculateUncollectedFees = func(positionLiquidity, feeGrowthBelow, feeGrowthAbove, feeGrowthInsideLast sdk.Dec, feeGrowthGlobal sdk.Dec) sdk.Dec {
+			// Calculating fee growth inside range [-1200; 400]
+			feeGrowthInside := calculateFeeGrowthInside(feeGrowthGlobal, feeGrowthBelow, feeGrowthAbove)
+
+			// Calculating uncollected fees
+			// Formula for finding uncollected fees in time range [t1; t2]:
+			// F_u = position_liquidity * (fee_growth_inside_t2 - fee_growth_inside_t1).
+			feesUncollected := positionLiquidity.Mul(feeGrowthInside.Sub(feeGrowthInsideLast))
+
+			return feesUncollected.TruncateDec()
+		}
+
+		// calculateSqrtPriceAtTick sqrtPrice at tickIndex
+		// View x/concentrated-liquidity/README.md -> ConcentratedLiquidity/Architecture/Ticks/Formulas for more details
+		calculateSqrtPriceAtTick = func(tickIndex, exponentAtPriceOne sdk.Int) sdk.Dec {
```

Comment:
Can we import this helper form concentrated liquidity? If it's private, I propose exporting it and moving to `types`